### PR TITLE
Introduce `PublicInstancePermission`

### PR DESCRIPTION
**Pull request by KaiVolland**
## Description

This introduces the `PublicInstancePermission`.

It allows to create a one to one relation to any BaseEntity via the `PublicInstancePermissionService` and the `BasePermissionController`.
If a `BaseEntity` is a `PublicInstancePermission` it is accesible for anyone, so no Authorization or User is required.

To be able to check for this permission it was needed to adapt the `DefaultWebSecurityConfig` and the `BaseCrudRepository`. 
:exclamation: As theses changes might be impactful this PR should be considered as a breaking and change. :exclamation: 

So it is marked with commit [90474d0](https://github.com/terrestris/shogun/pull/850/commits/90474d082b4c2c3c8a3a4483f8e06647cbe1c10f).

To set/unset an entity as public you can use the new `BasePermissionController` endpoint:

```
/[ENTITY]/[ID]/permissions/public
```

Just send an empty `POST` request to add the entity to the `publicinstancepermissions` table or a `DELETE` request to remove it.

This superseeds the approaches for the PRs of public applications #848 and public imagefiles #849 .

## Pull request type

- [ ] Bugfix
- [x] Feature
- [ ] Dependency updates
- [ ] Tests
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## Do you introduce a breaking change?

- [x] Yes
- [ ] No

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the 
  [Apache Licence Version 2.0](https://github.com/terrestris/shogun/blob/main/LICENSE).
- [x] I have followed the [guidelines for contributing](https://github.com/terrestris/shogun/blob/main/CONTRIBUTING.md).
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/terrestris/shogun/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have added or updated tests and documentation, and the test suite passes (run `mvn test` locally).
- [ ] I have added a screenshot/screencast to illustrate the visual output of my update.

### CODE REVIEW

File: shogun-config/src/main/java/de/terrestris/shogun/config/DefaultWebSecurityConfig.java [hunk 1]
```diff
@@ -43,7 +43,17 @@ default void customHttpConfiguration(HttpSecurity http) throws Exception {
...
+                    // Enable anonymous access to entity endpoints
+                    // will be secured via permission evaluators
+                    "/applications",
+                    "/applications/**",
+                    "/layers",
+                    "/layers/**",
+                    "/files/**",
+                    "/files",
+                    "/imagefiles/**",
+                    "/imagefiles"
```
Comment (dnlkoch) [inline shogun-config/src/main/java/de/terrestris/shogun/config/DefaultWebSecurityConfig.java:56]: Do we really need to allow non GET requests to anonymous users?

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/model/security/permission/PublicEntity.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+@Entity(name = "publicentities")
+@Table(schema = "shogun")
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString(callSuper = true)
+@EqualsAndHashCode
+public class PublicEntity implements Serializable {
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shogun/lib/model/security/permission/PublicEntity.java:36]: As this class refers to permissions I would recommend to rename it to something like `PublicInstancePermission`. So it would be more in sync with the name of the `/permissions` controller and the existing permission models.

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/security/access/entity/BaseEntityPermissionEvaluator.java [hunk 3]
```diff
@@ -75,6 +75,16 @@ public Class<E> getEntityClassName() {
     public boolean hasPermission(User user, E entity, PermissionType permission) {
         final String simpleClassName = entity.getClass().getSimpleName();
 
+        if (hasPublicPermission(entity)) {
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shogun/lib/security/access/entity/BaseEntityPermissionEvaluator.java:78]: Public access should be granted to `READ` permissions only. This should be checked here.

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/security/access/BasePermissionEvaluator.java [hunk 4]
```diff
@@ -45,13 +46,21 @@ public class BasePermissionEvaluator implements PermissionEvaluator {
...
     @Override
     public boolean hasPermission(Authentication authentication, Object targetDomainObject,
             Object permissionObject) {
 
         log.trace("About to evaluate permission for user '{}' targetDomainObject '{}' " +
             "and permissionObject '{}'", authentication, targetDomainObject, permissionObject);
 
+        if (hasPublicPermission(targetDomainObject)) {
+            return true;
+        }
```
Comment (dnlkoch) [inline shogun-lib/src/main/java/de/terrestris/shogun/lib/security/access/BasePermissionEvaluator.java:62]: Why is this needed? Isn't this check included in the `BaseEntityPermissionEvaluator`?

File: shogun-boot/src/main/resources/db/migration/V0.13.0__add_publicentities_table.sql [hunk 5]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS shogun.publicentities (
+    id bigint PRIMARY KEY,
+    created timestamp without time zone,
+    modified timestamp without time zone,
+	entity_id int8 NOT NULL UNIQUE
+);
```
Comment (dnlkoch) [inline shogun-boot/src/main/resources/db/migration/V0.13.0__add_publicentities_table.sql:6]: Do we need the audit tables as well?

File: shogun-config/src/main/java/de/terrestris/shogun/config/DefaultWebSecurityConfig.java (hunk 1 above)
Comment (KaiVolland) [reply shogun-config/src/main/java/de/terrestris/shogun/config/DefaultWebSecurityConfig.java:56]: I didn't know that it is possible to specifiy this by the method. I guess this would also makes sense for some of the other (mostly static resource) pathes here.

**Comment by KaiVolland** [comment]
Thanks for the review @dnlkoch. I gonna add some tests regarding the public topic.

File: shogun-lib/src/main/java/de/terrestris/shogun/lib/controller/security/permission/BasePermissionController.java [hunk 6]
```diff
@@ -753,6 +754,78 @@ public void deleteGroupClassPermissions(
...
+        try {
+            Optional<S> entity = service.findOne(entityId);
+
+            if (entity.isEmpty()) {
+                throw new EntityNotFoundException(entityId, getGenericClassName(), messageSource);
+            }
+
+            return Map.of("public", publicInstancePermissionService.getPublic(entity.get()));
+        } catch (AccessDeniedException ade) {
+            throw new EntityAccessDeniedException(entityId, getGenericClassName(), messageSource);
```
Comment (hwbllmnn) [inline shogun-lib/src/main/java/de/terrestris/shogun/lib/controller/security/permission/BasePermissionController.java:771]: I'd suggest to log the stack traces and error messages here and in the following catch clauses to enable better debugging

**Review by hwbllmnn — APPROVED** [review]

**Review by dnlkoch — APPROVED** [review]
